### PR TITLE
VCS: Port Godot 3.5's VCS features to GDExtension

### PR DESCRIPTION
**Pull request by twaritwaikar**
Pull the newer VCS features and an improved diff viewer from 3.5 to 4

Watch this demo for this intended usage - https://youtu.be/blVtgs4g_GU

I was able to port ALL of the features as they exist in 3.5. The next task is to pull out the UI into the plugin as per https://github.com/godotengine/godot-proposals/issues/4369, but in case that turns out to be too difficult, getting this PR merged will ensure the VCS integration works as expected.

### Issues

I tried some fixes to clean the implementation up but there are still some issues that don't seem to be related to the changes in this PR but they can be put into newer GH issues:

* If a script has been opened in the code editor, and you try to write a commit message, currently, the code editor will _steal_ focus from the commit message textbox on every keystroke, making it impossible to write a proper commit message. I have to switch to the 3D tab or the 2D tab to write a commit message.

image
image

* `RichTextLabel::append_text()` is not making the text `[center]`'d properly. The text surrounded by `@@` should have been centered.
* The `push_indent` within `RichTextLabel`'s table cells is not working as expected in the split view. The line numbers are overlapping with the content. This doesn't appear to happen in the unified view.

### CODE REVIEW

**Comment by ChronicallySerious** [comment]
Currently godot-cpp is crashing after my recent rebase on master. Reported https://github.com/godotengine/godot-cpp/issues/800

**Comment by ChronicallySerious** [comment]
I have added a demo of the changes in my PR description

File: editor/editor_vcs_interface.cpp [hunk 1]
```diff
@@ -30,132 +30,371 @@
 
 #include "editor_vcs_interface.h"
 
+#include "editor_node.h"
+
+#define UNIMPLEMENTED() ERR_PRINT(TTR("Unimplemented virtual function in EditorVCSInterface based node: ") + __func__)
```
Comment (akien-mga) [inline editor/editor_vcs_interface.cpp]: We normally don't translate console errors, so this shouldn't use `TTR`.
For errors displayed in the editor (e.g. as you do below with `EditorNode::show_warning()`) you should use `TTR` indeed.

File: editor/plugins/version_control_editor_plugin.cpp [hunk 2]
```diff
@@ -30,47 +30,81 @@
 
 #include "version_control_editor_plugin.h"
 
-#include "core/object/script_language.h"
+#include "core/config/project_settings.h"
+#include "core/core_bind.h"
```
Comment (akien-mga) [inline editor/plugins/version_control_editor_plugin.cpp:34]: I couldn't spot why it's included in the file but in general `core/core_bind.h` shouldn't be used outside core bindings. It's only meant to wrap some core functionality in a way that's convenient for GDScript / other scripting languages. But for C++ you should use the underlying core functionality directly.

File: editor/plugins/version_control_editor_plugin.cpp [hunk 3]
```diff
@@ -141,8 +143,8 @@ void VersionControlEditorPlugin::_initialize_vcs() {
 	if (_load_plugin(selected_plugin)) {
 		_set_up();
 
-		ProjectSettings::get_singleton()->set("editor/version_control/autoload_on_startup", true);
-		ProjectSettings::get_singleton()->set("editor/version_control/plugin_name", selected_plugin);
+		ProjectSettings::get_singleton()->set(SNAME("editor/version_control/autoload_on_startup"), true);
+		ProjectSettings::get_singleton()->set(SNAME("editor/version_control/plugin_name"), selected_plugin);
```
Comment (akien-mga) [inline editor/plugins/version_control_editor_plugin.cpp:149]: This doesn't need to use `SNAME()`. `SNAME()` allocates static memory and should only be used in critical code paths which are used a lot. I assume this one only runs once so it's fine to let it convert the strings to StringName on the fly.

File: editor/plugins/version_control_editor_plugin.cpp [hunk 4]
```diff
@@ -169,34 +171,27 @@ void VersionControlEditorPlugin::_set_credentials() {
...
-	ScriptInstance *plugin_script_instance = script->instance_create(vcs_interface);
-
-	ERR_FAIL_COND_V_MSG(!plugin_script_instance, false, "Failed to create plugin script instance.");
-
-	// The plugin is attached as a script to the VCS interface singleton as a proxy end-point
-	vcs_interface->set_script_instance(plugin_script_instance);
-
-	EditorVCSInterface::set_singleton(vcs_interface);
+	EditorVCSInterface *vcs_plugin = Object::cast_to<EditorVCSInterface>(extension_instance);
+	ERR_FAIL_NULL_V_MSG(vcs_plugin, false, vformat(TTR("Could not cast VCS extension instance to %s."), EditorVCSInterface::get_class_static()));
```
Comment (akien-mga) [inline editor/plugins/version_control_editor_plugin.cpp:189]: I won't comment on all of them but you should remove all `TTR()` in `ERR_FAIL.*` macros.

As you can see before this PR there's almost no code translating `ERR_FAIL` errors:
```
$ rg ERR_FAIL.*TTR
editor/editor_vcs_interface.cpp
173:                    ERR_FAIL_MSG(TTR("Couldn't create .gitignore in project path."));
180:                    ERR_FAIL_MSG(TTR("Couldn't create .gitattributes in project path."));

editor/import/resource_importer_imagefont.cpp
115:    ERR_FAIL_COND_V_MSG(err != OK, ERR_FILE_CANT_READ, TTR("Can't load font texture:") + " \"" + p_source_file + "\".");
```

**Review by akien-mga — APPROVED** [review]
This is a pretty big PR so not trivial to review, but overall it seems well made, and pretty self-contained so there should be no major problem to merge it.

**Comment by ChronicallySerious** [comment]
Also, @groud had reviewed some of the code that touches things outside of the VCS on [chat](https://chat.godotengine.org/channel/editor?msg=K9tXbkXwXN6yJ692n). There seems to be another bug where the CodeEditor is not updating the script files if they get changed from outside the editor (at least on Windows). I will open a separate bug to track this

**Comment by ChronicallySerious** [comment]
Once that is solved, we may not need to call `ScriptEditor::reload_scripts()` through this code. There needs to be additional testing on the behavior in that case.

**Comment by akien-mga** [comment]
Might need a rebase after #64377 I think.

**Comment by ChronicallySerious** [comment]
@akien-mga Rebased, and replaced `update()` with `queue_redraw()`

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**